refactor (#325)
KxSystems/docs: docs/architecture/tickq.md

---
title: tick.q | Documentation for q and kdb+
description: How to construct a tickerplant process
keywords: hdb, kdb+, q, tick, tickerplant, streaming
---
# Tickerplant (TP) using tick.q

`tick.q` is available from :fontawesome-brands-github:[KxSystems/kdb-tick](https://github.com/KxSystems/kdb-tick)

## Overview

All incoming streaming data is processed by a kdb+ process acting as a tickerplant.
A tickerplant writes all data to a tickerplant log (to permit data recovery) and publishes data to subscribed clients, for example a RDB.

### Customization

> `tick.q` provides a starting point to most environments. The source code is freely avaialble and can be tailered to individual needs.

Reviewer: ```suggestion
`tick.q` provides a starting point to most environments. The source code is freely available and can be tailored to individual needs.
```

### Schema file

A tickerplant requires a schema file.

> A schema file should be created to describe the data you plan to capture, by specifing the tables that will be populated by the tickerplant environment. 

Reviewer: ```suggestion
A schema file describes the data you plan to capture, by specifying the tables to be populated by the tickerplant environment. 
```

The [datatypes](../basics/datatypes.md) and [attributes](../ref/set-attribute.md) are denoted within the file as shown in this example:
```q
quote:([]time:`timespan$(); sym:`g#`symbol$(); bid:`float$(); ask:`float$(); bsize:`long$(); asize:`long$(); mode:`char$(); ex:`char$())
trade:([]time:`timespan$(); sym:`g#`symbol$(); price:`float$(); size:`int$(); side:`char$())
```

The default setup requires the first two columns to be `time` and `sym`.

### Real-time vs Batch Mode

_The mode is controlled via the [`-t`](#usage) command line parameter._
Batch mode can alleviate CPU use on both the tickerplant and its subscribers by grouping together multiple ticks within the timer interval prior to sending/writing. 

> This will be at the expense of tickerplant memory (required memory to hold several ticks) and increased latency that may occur between adding to the batch and sending. 

Reviewer: ```suggestion
This comes at the expense of tickerplant memory (required memory to hold several ticks) and increased latency that may occur between adding to the batch and sending. 
```

There is no ideal setting for all deployments as it depends on the frequency of the ticks received. 

> Real-time mode will process every tick as soon as they occur.

Reviewer: ```suggestion
Real-time mode processes every tick as soon as they occur.
```

* call [`.u.init[]`](uq.md#uinit) to initialise table info, [`.u.t`](#variables) and [`.u.w`](#variables)
* check first two columns in all tables of provided schema are called `time` and `sym` (throw `timesym` error if not)
* apply [`grouped`](../ref/set-attribute.md#grouped-and-parted) attribute to the sym column of all tables in provided schema
* set [`.u.d`](#variables) to current local date, using [`.z.D`](../ref/dotz.md#zt-zt-zd-zd-timedate-shortcuts)
* if a tickerplant log filename was provided:
    * set [`.u.L`](#variables) with a temporary value of `` `:<log filename>/<schema filename>.......... `` (will have date added in next step)
    * create/initialise the log file by calling [`.u.ld`](#uld), passing [`.u.d`](#variables) (current local date)
    * set [`.u.l`](#variables) to log file handle

### .u.ld

Initialise or reopen existing log file.

```q
.u.ld[x]
```

Where `x` is current date. Returns handle of log file for that date.

Actions performed:

* using [`.u.L`](#variables), change last 10 chars to provided date and create log file if it doesnt yet exist
* set [`.u.i`](#variables) and [`.u.j`](#variables) to count of valid messages currently in log file

> * if log file is found to be corrupt (size bigger than size of number of valid messages) an error will be returned

Reviewer: ```suggestion
* if log file is found to be corrupt (size bigger than size of number of valid messages) an error is returned
```

* open new/existing log file

### .u.ts

Given a date, runs end-of-day procedure if a new day has started.

```q
.u.ts[x]
```
Where x is a date.

Compares date provided with [`.u.d`](#variables). If no change, no action taken. 
If one day difference (i.e. a new day), [`.u.endofday`](#uendofday) is called. 
More than one day results in an error and the kdb+ timer is cancelled.

### .u.upd

Update tickerplant with data to process/analyse. External processes call this to input data into the tickerplant.

```q
.u.upd[x;y]
```
Where

* `x` is table name (sym)
* `y` is data for table `x` (list of column data, each element can be an atom or list)

#### Batch Mode

> Add each recieved message to batch and record message to tickerplant log. Batch will be published on running timer.

Reviewer: ```suggestion
Add each received message to the batch and record message to the tickerplant log. Batch is published on running timer.
```
